### DISCUSSION TITLE
[gatsby-source-wordpress] warn ./node_modules/gatsby-source-wordpress/dist/gatsby-browser.js Attempted import error: 'createRoot' is not exported from 'react-dom' (imported as 'ReactDOM').

### DISCUSSION THREAD

**Discussion by qpiddigital** [gatsbyjs/gatsby#33856] [Help]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

After upgrading to Gatsby 4, I get a warning message when running npm run develop:

warn ./node_modules/gatsby-source-wordpress/dist/gatsby-browser.js Attempted import error: 'createRoot' is not exported from 'react-dom' (imported as 'ReactDOM').

Also, I've noticed when the development server is running and "watching for WordPress changes", it doesn't pick up changes after I update pages and posts in WordPress. Not sure if this is related.

### Reproduction Link

n/a

### Steps to Reproduce

1. Upgraded to Gatsby 4
2. npm run develop

### Expected Result

n/a

### Actual Result

n/a

### Environment

```shell
System:
    OS: macOS 11.6
    CPU: (16) x64 Intel(R) Xeon(R) W-2140B CPU @ 3.20GHz
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 17.0.1 - /usr/local/bin/node
    npm: 8.1.0 - /usr/local/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Chrome: 95.0.4638.69
    Safari: 15.0
  npmPackages:
    gatsby: ^4.0.2 => 4.1.0
    gatsby-plugin-emotion: ^7.0.0 => 7.1.0
    gatsby-plugin-image: ^2.0.0 => 2.1.0
    gatsby-plugin-manifest: ^4.0.0 => 4.1.0
    gatsby-plugin-postcss: ^5.0.0 => 5.1.0
    gatsby-plugin-react-helmet: ^5.0.0 => 5.1.0
    gatsby-plugin-sharp: ^4.0.1 => 4.1.0
    gatsby-source-filesystem: ^4.0.0 => 4.1.0
    gatsby-source-wordpress: ^6.0.0 => 6.1.0
    gatsby-transformer-sharp: ^4.0.0 => 4.1.0
```

**Comment by LekoArts** [comment]
Your npm module tree might be messed up or you have multiple versions of React installed. Try a `gatsby clean`, then delete the `node_modules` folder and `package-lock.json` and run `npm install --legacy-peer-deps`.

**Reply by RectoVersoDev** [reply]
Hi, I did this but the warning is still here :(

Edit: Same warning with fresh install `gatsby-starter-default` (+ `gatsby-source-wordpress`)

> System:
    OS: Windows 10 10.0.19043
    CPU: (12) x64 AMD Ryzen 5 3600 6-Core Processor
  Binaries:
    Node: 14.17.4 - C:\Program Files\nodejs\node.EXE
    npm: 7.21.1 - C:\Program Files\nodejs\npm.CMD
  Languages:
    Python: 3.10.0
  Browsers:
    Chrome: 95.0.4638.69
    Edge: Spartan (44.19041.1266.0), Chromium (95.0.1020.40)
  npmPackages:
    gatsby: ^4.1.0 => 4.1.0
    gatsby-omni-font-loader: ^1.3.3 => 1.3.3
    gatsby-plugin-image: ^2.1.0 => 2.1.0
    gatsby-plugin-manifest: ^4.1.0 => 4.1.0
    gatsby-plugin-no-index: ^1.0.2 => 1.0.2
    gatsby-plugin-react-helmet: ^5.1.0 => 5.1.0
    gatsby-plugin-sharp: ^4.1.0 => 4.1.0
    gatsby-plugin-styled-components: ^5.1.0 => 5.1.0
    gatsby-source-filesystem: ^4.1.0 => 4.1.0
    gatsby-source-wordpress: ^6.1.0 => 6.1.0
    gatsby-transformer-sharp: ^4.1.0 => 4.1.0
  npmGlobalPackages:
    gatsby-cli: 4.1.0

**Reply by CesarBenavides777** [reply]
@LekoArts this also didn't work for me.

**Comment by miraclemaker** [comment]
I'm getting the same error message on a brand new install via "gatsby new"

  System:
    OS: Windows 10 10.0.18363
    CPU: (16) x64 AMD Ryzen 7 1700 Eight-Core Processor
  Binaries:
    Node: 16.13.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.5 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 7.24.2 - C:\Program Files\nodejs\npm.CMD
  Languages:
    Python: 3.10.0 - /c/Python310/python
  Browsers:
    Chrome: 95.0.4638.69
    Edge: Spartan (44.18362.1533.0)
  npmPackages:
    gatsby: ^4.1.0 => 4.1.0
    gatsby-plugin-image: ^2.1.0 => 2.1.0
    gatsby-plugin-react-helmet: ^5.1.0 => 5.1.0
    gatsby-plugin-sass: ^5.1.0 => 5.1.0
    gatsby-plugin-sharp: ^4.1.0 => 4.1.0
    gatsby-plugin-sitemap: ^5.1.0 => 5.1.0
    gatsby-source-filesystem: ^4.1.0 => 4.1.0
    gatsby-source-wordpress: ^6.1.0 => 6.1.0
    gatsby-transformer-sharp: ^4.1.0 => 4.1.0
  npmGlobalPackages:
    gatsby-cli: 4.1.0

> warn ./node_modules/gatsby-source-wordpress/dist/gatsby-browser.js
> Attempted import error: 'createRoot' is not exported from 'react-dom' (imported as 'ReactDOM').

**Reply by CesarBenavides777** [reply]
Same. Ever since I updated to Gatsby v4 this issue is persistant across all sites. Seems to break the lighthouse scores on Gatsby Cloud.

**Reply by b-sik** [reply]
Experiencing the same problem on MacOS after upgrading to Gatsby 4. I migrated to `yarn` to resolve another issue, but this issue is persisting.

**Comment by HenrijsS** [comment]
Created a new Gatsby project for the first time ever and receive this warning every time I create a change in my code. 
It's reeeealy annoying, but so far doesn't break anything.

**Reply by b-sik** [reply]
It looks like `gatsby-source-wordpress` is using an [experimental React 18 function `createRoot`](https://reactjs.org/docs/concurrent-mode-reference.html). I removed the references of `createRoot` in `../node_modules/gatsby-source-wordpress/dist/gatsby-browser.js` on my local project and replaced with `render` and the error goes away and everything seems to work fine.

Interestingly, there's comments in the file saying `// @ts-ignore - createRoot is on ReactDOM`

**Reply by CesarBenavides777** [reply]
My gatsby wordpress also doesn't seem to fetch changes from wordpress anymore I'm having to clear .wordpress-cache to see new changes. Might be related.

**Reply by b-sik** [reply]
Made a PR with my solution above - https://github.com/gatsbyjs/gatsby/pull/33862

**Reply by Haldaug** [reply]
I have the same issue as @CesarBenavides777 with gatsby-plugin-wordpress not fetching changes from wordpress anymore. I also see the same error message.

I tried @bszyk's solution and the error message went away, but gatsby-source-worpdress is still not fetching changes.

**Reply by kindredesign** [reply]
@Haldaug @CesarBenavides777  Unfortunately, we're having exactly the same issue as you both where we have to clean cache for changes to update. 

Have noticed a new version of WpGatsby (v2.0.1) released 2 days ago, I wonder if the issue could relate to this?

**Reply by Haldaug** [reply]
@kindredesign At least there are more people having this problem, making it easier to troubleshoot it. I think I'll open an issue.

**Reply by Haldaug** [reply]
@CesarBenavides777 @kindredesign I tried reverting the WPGatsby plugin to version 1.1.4, but that didn't help.

**Reply by CesarBenavides777** [reply]
Yeah @Haldaug also tried with version `2.0.0` no go.

**Reply by kindredesign** [reply]
@CesarBenavides777  @Haldaug I've just switched the Wordpress backend to a staging version we were using... and it appears to be working again. Live updates pulled in development mode (and I suspect this would also work fine when building although haven't tested it).

Suggests something on WP side. I'll do a comparison between the two Wordpress environments and see if I can spot an issue!

**Edit:** Working (staging version) appears to be using WPGatsby 1.1.3.

**Reply by HenrijsS** [reply]
@kindredesign So the issue is in the core WordPress plugin and not on Gatsby side?

EDIT: If so, I suggest rolling back the plugin.
Here's a quick way to do it: https://wordpress.org/plugins/wp-rollback/

**Reply by HenrijsS** [reply]
Adding a new message:

The update error is only on WP **!GraphQL!** 1.6.7. Using the rollback plugin I provided, downgrade to 1.6.6.
Seems to be a new change there that breaks something.

**Reply by CesarBenavides777** [reply]
@HenrijsS confirmed that works

**Reply by HenrijsS** [reply]
I created issues in both WPGrahql and WPGatsby repositories. Don't know which side breaks it, so for now just downgrade **WPGraphQL** to 1.6.6. 
WPGrahql: https://github.com/wp-graphql/wp-graphql/issues/2142
WPGatsby: https://github.com/gatsbyjs/wp-gatsby/issues/200

**Reply by kindredesign** [reply]
Can confirm rolling WPGraphQL back to 1.6.6 solves the issue from my end too.

Thanks @HenrijsS @CesarBenavides777 @Haldaug. A good few hours spent last night and this morning looking to solve that one!!

**Reply by Haldaug** [reply]
A downgrade of WPGraphQL to 1.6.6 solved it for me too. 

Thanks for the communal troubleshooting, @HenrijsS @CesarBenavides777  and @kindredesign!

**Reply by TylerBarnes** [reply]
This issue was separate from the original issue in the title of this discussion but I fixed the data update problem in WPGatsby `v2.0.2` just now.

Note that WPGatsby `v2.0.0` is a breaking change as how Gatsby Previews work has been radically changed. If you can't upgrade WPGatsby yet, roll back to WPGraphQL `v1.6.6`. Ideally WPGraphQL would've made a breaking release for the change they put out in `v1.6.7`.

**Comment by LekoArts** [comment]
So this appears to be specific to WordPress. Please follow https://github.com/wp-graphql/wp-graphql/issues/2142 & https://github.com/gatsbyjs/wp-gatsby/issues/200.

On a side-note: This discussion here shows why a minimal reproduction and filling out our issue template is important. It would have shown that this only happens in combination with WordPress and the latest versions earlier. Makes it easier for both sides.

**Reply by Haldaug** [reply]
I wouldn't say that the original issue is specific to WordPress and the wp-graphql plugin. The issues you are referring to are specific to the problem that gatsby-source-wordpress wasn't fetching changes. This was found to be unrelated to the original issue about the "Attempted import error" message.

**Reply by TylerBarnes** [reply]
Yeah, it looks like there are 2 separate issues here.

1. A potentially breaking change in WPGraphQL
2. Usage of createRoot here is somehow causing problems https://github.com/gatsbyjs/gatsby/pull/33138/files#diff-19903ad5f673d6e63bbaf02d1aad9ccb98ceae8adb70ffb9b87be6fbcbb3427fR54

**Reply by Haldaug** [reply]
We are following up issue number 1 over here: [https://github.com/wp-graphql/wp-graphql/issues/2142](https://github.com/wp-graphql/wp-graphql/issues/2142)

**Reply by TylerBarnes** [reply]
For issue 1 I fixed the data update problem in WPGatsby `v2.0.2` just now.

Note that WPGatsby `v2.0.0` is a breaking change as how Gatsby Previews work has been radically changed. If you can't upgrade WPGatsby yet, roll back to WPGraphQL `v1.6.6`. Ideally WPGraphQL would've made a breaking release for the change they put out in `v1.6.7`.

**Reply by xy-lin** [reply]
Problem solved after upgrading to WPGatsby 2.0.2 and WPGraphQL 1.6.7.
Many thanks for effort and time.

**Reply by RectoVersoDev** [reply]
Issue 2 is solved for everyone ? I still get the warning..

`warn ./node_modules/gatsby-source-wordpress/dist/gatsby-browser.js Attempted import error: 'createRoot' is not exported from 'react-dom' (imported as 'ReactDOM').`

gatsby 4.1.4
gatsby-source-wordpress 6.1.2
WP Gatsby 2.0.2
WP GraphQL 1.6.7

Changes detection is fixed tho

**Reply by CesarBenavides777** [reply]
@RectoVersoDev still experiencing this:
![image](https://user-images.githubusercontent.com/17585202/141798517-1ea39ec4-9a7d-4762-a6e1-cffb6bd0ec6b.png)

**Reply by miraclemaker** [reply]
Yes still getting the issue here too.

**Reply by qpiddigital** [reply]
Same issue here too.

**Reply by elliott-w** [reply]
Still getting the issue

**Comment by wardpeet** [comment]
We'll publish a fix for this warning in gatsby-source-wordpress@6.2.0 (will be out later today)

**Reply by qpiddigital** [reply]
Confirmed - no longer receiving warning message after upgrading to gatsby-source-wordpress@6.2.0. Great work, thank you.

**Reply by RectoVersoDev** [reply]
Same, no more warning message, thank you guys

**Reply by gvazco** [reply]
I confirm that this is the correct solution!
